src/Credentials/GCECredentials.php

```diff
@@ -340,6 +345,22 @@ public static function onGce(callable $httpHandler = null)
             } catch (ConnectException $e) {
             }
         }
+
+        if (PHP_OS === 'Windows') {
+            // @TODO: implement GCE residency detection on Windows
+            return false;
+        }
+
+        // Detect GCE residency on Linux
+        return self::detectResidencyLinux(self::GKE_PRODUCT_NAME_FILE);
+    }
+
+    private static function detectResidencyLinux(string $productNameFile): bool
+    {
+        if (file_exists($productNameFile)) {
+            $productName = trim((string) file_get_contents($productNameFile));
+            return 0 === strpos($productName, 'Google');
```

Comment:
Maybe lowercase is better?
```suggestion
            return 0 === strpos($productName, 'google');
```

Reply:
My only question is: Should we normalize the contents (e.g. force lowercase) in order to avoid any sort of subtle issues with string matching/discrepancies in this file's contents?

Reply:
The Product Name is either `Google` or `Google Compute Engine`. The lowercase name `google` is not valid (or at least, there are no known Operating Systems which use that). So there's no reason to add a check for the lowercase value.

To my knowledge, other language implementations have not included such a check